lib/jwt/jwk/key_base.rb

```diff
@@ -8,28 +8,34 @@ def self.inherited(klass)
         ::JWT::JWK.classes << klass
       end
 
-      def initialize(options)
+      def initialize(options, params = {})
         options ||= {}
 
-        if options.is_a?(String) # For backwards compatibility when kid was a String
-          options = { kid: options }
-        end
+        @parameters = params.transform_keys(&:to_sym) # Uniform interface
 
-        @kid           = options[:kid]
-        @kid_generator = options[:kid_generator] || ::JWT.configuration.jwk.kid_generator
+        # For backwards compatibility, kid_generator may be specified in the parameters
+        options[:kid_generator] ||= @parameters.delete(:kid_generator)
```

Comment:
The `kid_generator` value will not be deleted from `@parameters` if `options` contains the `kid_generator`

Reply:
This was kinda intended, though poorly implemented.
The situation we are worried about there involves someone specifying `kid_generator` in both `options` and `params`.
Think about different kinds of application programmers:

One kind of programmer is using the API in its current form. They will only specify `kid_generator` in the parameters, since currently, there is no third argument for `initialize`.

The second kind of programmer will use the new JWK API additions. The new README says that `kid_generator` needs to be specified in the third argument. And if they happen to look at the code, they will see the words `backwards compatibility`, and the handling right below.

This latter kind of programmer will probably also expect that anything in the `params` argument (which is not kty-related) will stay a part of the JWK. In some rare circumstances, this will include `kid_generator` for some obscure application (Do not worry, I _will_ quote you when that happens ;D). By specifying the `kid_generator` explicitly in the `options`, they have a workaround if this is indeed their expected behavior.

I do not expect any other kinds of programmers to actively make use of the option to specify `kid_generator` twice.
What do you think?

Reply:
Everything makes sense, just leave it like this. It's simple